mlir/lib/Analysis/BufferDependencyAnalysis.cpp

```diff
@@ -0,0 +1,153 @@
+//===- BufferDependencyAnalysis.h - a set of useful generic analyses
+//---------------===//
+//
+// Part of the MLIR Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Analysis/BufferDependencyAnalysis.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+#include <optional>
+
+using namespace mlir;
+
+struct UserWrapper {
+  Operation *parentAlias;
+  Operation *op;
+};
+
+void findAllUsers(Operation *parent, llvm::SmallVector<UserWrapper> &users) {
+  for (auto *user : parent->getUsers()) {
+    if (auto viewOp = dyn_cast<ViewLikeOpInterface>(user)) {
+      findAllUsers(viewOp, users);
+    } else {
+      users.push_back({parent, user});
+    }
+  }
+}
+
+enum EffectType { read, write, unknown };
+EffectType getMemoryEffect(UserWrapper &user) {
```

Comment:
It's a two pointer struct, might as well pass it by value, especially since you're not modifying it?